### PR TITLE
gamma: change API for mapping input to region argument

### PR DESCRIPTION
**Pull request by caleridas**
Remove the GammaInput class and the way it is used to map an input to the corresponding argument inside the lambda region. Replace this with an API function attached to the node instead.

### CODE REVIEW

File: jlm/rvsdg/gamma.cpp [hunk 1]
```diff
@@ -265,79 +265,130 @@ GammaOperation::operator==(const operation & other) const noexcept
...
-  size_t n;
-  auto origin = argument->input()->origin();
-  for (n = 1; n < nresults(); n++)
+  return ev;
+}
+
+GammaNode::EntryVar
+GammaNode::GetEntryVar(std::size_t index) const
+{
+  EntryVar ev;
```
Comment (phate) [inline jlm/rvsdg/gamma.cpp:299]: Could we get an assert() checking that the index is within bounds?

File: jlm/rvsdg/gamma.cpp [hunk 2]
```diff
@@ -265,79 +265,130 @@ GammaOperation::operator==(const operation & other) const noexcept
...
+  if (values.size() != nsubregions())
+    throw jlm::util::error("Incorrect number of values.");
 
-  auto node = input_->node();
-  auto index = input_->index();
-  if (index == node->ninputs() - 1)
+  const auto & type = values[0]->Type();
+  auto output = node::add_output(std::make_unique<rvsdg::StructuralOutput>(this, type));
+
+  std::vector<rvsdg::input *> branch_results;
```
Comment (phate) [inline jlm/rvsdg/gamma.cpp:343]: camelCase

File: jlm/rvsdg/gamma.cpp [hunk 3]
```diff
@@ -265,79 +265,130 @@ GammaOperation::operator==(const operation & other) const noexcept
...
+  for (size_t n = 0; n < nsubregions(); n++)
   {
-    input_ = nullptr;
-    return *this;
+    branch_results.push_back(&rvsdg::RegionResult::Create(
+        *subregion(n),
+        *values[n],
+        static_cast<rvsdg::StructuralOutput *>(output),
```
Comment (phate) [inline jlm/rvsdg/gamma.cpp:349]: Move the cast up to the creation of output.

File: jlm/rvsdg/gamma.cpp [hunk 4]
```diff
@@ -265,79 +265,130 @@ GammaOperation::operator==(const operation & other) const noexcept
...
-GammaNode::exitvar_iterator::operator++() noexcept
+std::vector<GammaNode::ExitVar>
+GammaNode::GetExitVars() const
 {
-  if (output_ == nullptr)
-    return *this;
+  std::vector<GammaNode::ExitVar> vars;
+  for (size_t n = 0; n < noutputs(); ++n)
+  {
+    std::vector<rvsdg::input *> branch_results;
```
Comment (phate) [inline jlm/rvsdg/gamma.cpp:362]: camelCase

File: jlm/rvsdg/gamma.cpp [hunk 5]
```diff
@@ -265,79 +265,130 @@ GammaOperation::operator==(const operation & other) const noexcept
...
+  return vars;
+}
 
-  auto node = output_->node();
-  auto index = output_->index();
-  if (index == node->nexitvars() - 1)
+GammaNode::ExitVar
+GammaNode::MapOutputExitVar(const rvsdg::output & output) const
+{
+  std::vector<rvsdg::input *> branch_results;
```
Comment (phate) [inline jlm/rvsdg/gamma.cpp:375]: camelCase

Comment (phate) [inline jlm/rvsdg/gamma.cpp:375]: Some assert that output is the output of a gamma node.

File: jlm/rvsdg/gamma.cpp [hunk 6]
```diff
@@ -265,79 +265,130 @@ GammaOperation::operator==(const operation & other) const noexcept
...
   }
+  return ExitVar{ std::move(branch_results), node::output(output.index()) };
+}
 
-  output_ = node->exitvar(++index);
-  return *this;
+GammaNode::ExitVar
+GammaNode::MapBranchResultExitVar(const rvsdg::input & input) const
+{
+  std::vector<rvsdg::input *> branch_results;
```
Comment (phate) [inline jlm/rvsdg/gamma.cpp:386]: Some assert that the input is a result of the region

Comment (phate) [inline jlm/rvsdg/gamma.cpp:386]: camelCase

File: jlm/mlir/frontend/MlirToJlmConverter.cpp [hunk 7]
```diff
@@ -355,7 +355,7 @@ MlirToJlmConverter::ConvertOperation(
         JLM_ASSERT(regionResults[regionIndex].size() == regionResults[0].size());
         exitvars.push_back(regionResults[regionIndex][exitvarIndex]);
       }
-      rvsdgGammaNode->add_exitvar(exitvars);
+      rvsdgGammaNode->AddExitVar(exitvars).output;
```
Comment (phate) [inline jlm/mlir/frontend/MlirToJlmConverter.cpp:358]: The `.output` is dead code, no?

File: jlm/rvsdg/gamma.hpp [hunk 8]
```diff
@@ -117,180 +116,184 @@ class GammaNode : public StructuralNode
...
+  struct EntryVar
+  {
+    /**
+     * \brief Variable at entry point (input to gamma node).
+     */
+    rvsdg::input * input;
+    /**
+     * \brief Variable inside each of the branch regions (argument in per subregion).
+     */
+    std::vector<rvsdg::output *> branches;
```
Comment (phate) [inline jlm/rvsdg/gamma.hpp:132]: I really do not like the name `branches`. I find it very misleading. What about `subregionArgument` or even `branchArgument`?

File: jlm/rvsdg/gamma.hpp [hunk 9]
```diff
@@ -117,180 +116,184 @@ class GammaNode : public StructuralNode
...
-    GammaOutput * output_;
+  /**
+   * \brief A variable routed out of all gamma regions as result.
+   */
+  struct ExitVar
+  {
+    /**
+     * \brief Variable exit points (results per subregion).
+     */
+    std::vector<rvsdg::input *> branches;
```
Comment (phate) [inline jlm/rvsdg/gamma.hpp:143]: Same here.

File: jlm/rvsdg/gamma.cpp [hunk 10]
```diff
@@ -265,79 +265,130 @@ GammaOperation::operator==(const operation & other) const noexcept
   return op && op->nalternatives_ == nalternatives_;
 }
 
-/* gamma input */
-
-GammaInput::~GammaInput() noexcept = default;
+/* gamma node */
```
Comment (phate) [inline jlm/rvsdg/gamma.cpp:268]: Since you are already touching it, it can be removed....it adds no value.

**Review by phate — CHANGES_REQUESTED** [review]

File: jlm/rvsdg/gamma.hpp (hunk 8 above)
Comment (phate) [reply jlm/rvsdg/gamma.hpp:132]: Or any other ideas? If everything fails, we can use `branches`, but I am not in favor or it.

File: jlm/mlir/frontend/MlirToJlmConverter.cpp (hunk 7 above)
Comment (phate) [reply jlm/mlir/frontend/MlirToJlmConverter.cpp:358]: The CI complains about it as well.

File: jlm/rvsdg/gamma.hpp (hunk 8 above)
Comment (caleridas) [reply jlm/rvsdg/gamma.hpp:132]: Some names that I came up with

- branchVar (this is the vesion of the variable in the respective branches)
- branchVariable, same as branchVar just longer
- branchDef (based on eventually renaming output->Def and input->Use)
- branchArg, shorter than branchArgument
- perBranch: actually quite accurately describes what it is, but at the same time sounds awkward

initially I had some dislike for the term "argument", but actually on thinking more about it "branchArg(ument)" (and "branchRes(ult)" for the converse sounds suitable -- generally I prefer not to abbreviate and would use the full names, but not certain. WDYT?

Comment (caleridas) [reply jlm/rvsdg/gamma.hpp]: changed it branchArgument & branchResult, I think that is appropriate

Comment (phate) [reply jlm/rvsdg/gamma.hpp]: branchArgument and branchResult sounds great.

**Review by phate — DISMISSED** [review]
There are still two open issues from the last review round. I will not stop this PR on them though.

File: jlm/rvsdg/gamma.cpp (hunk 5 above)
Comment (caleridas) [reply jlm/rvsdg/gamma.cpp]: added assertion

**Comment by caleridas** [comment]
> There are still two open issues from the last review round. I will not stop this PR on them though.

I think I just addressed the two open issues while you were posting this -- I think you need to re-approve.

**Review by phate — APPROVED** [review]

**Pull request merged**